### DISCUSSION TITLE
Sharing our FreshMVVM registered dependencies with Shiny Delegates

### DISCUSSION THREAD

**Discussion by munkii** [shinyorg/shiny#1117] [Q&A]
We are using FreshMVVM and it's IOC Container which is using TinyIOC "under the covers". 

In the past when I have needed some Shiny classes injected into my FreshMVVM registered ViewModels I have done the following

`FreshMvvm.FreshIOC.Container.Register<INotificationManager>(ShinyHost.Resolve<INotificationManager>());`

That's worked just fine and now we need the opposite. Access to some of our services within some Shiny delegates such as IPushDelegate. The approach I have taken is to expolicitly call Resolve on the FreshIOC container

`var breatheServices = FreshIOC.Container.Resolve<IBreatheServices>();
var clinicalService = FreshIOC.Container.Resolve<IClinicalService>();`

Under normal debugging and test scenarios that seems to work but recently I wired up Shiny Logging for AppCenter via the `ShinyStartup.ConfigureLogging` and that highlighted that sometimes those Resolve calls are failing. This makes me think I should be configure the dependencies needed in the Shiny delegates using the correct Shiny process.

I added some AddSingleton registration in ShinyStartup.ConfigureServices

public override void ConfigureServices(IServiceCollection services, IPlatform platform)
{
            services.UsePush<PushDelegate>();
            services.UseNotifications<NotificationDelegate>();

            services.AddSingleton<IDeviceInformationService, DeviceInformationService>();
            services.AddSingleton<INetworkInformationService, NetworkInformationService>();
            services.AddSingleton<IClinicalService, ClinicalService>();
            services.AddSingleton<IDataProviderService, DataProviderService>();
            services.AddSingleton<IBreatheServices, BreatheServices>();
}

And then added the dependencies to the PushDelagate ctor. Now when I run the app I get no errors but the PushDelegate seems to be invalid with none of the breakpoints firing and no System.Diagnostics.Debug.WriteLine outputs appearing in the output window

Have I totally misunderstood how I should get my dependencies into the likes of the PushDelegate?
Is there a way of reusing the FreshMVVM IOC container with Shiny?
We are using FreshMVVM and it's IOC Container which i believe is using TinyIOC. 

In the past when I have needed access to some Shiny classes injected into my FreshMVVM registered ViewModels I have done the following

`FreshMvvm.FreshIOC.Container.Register<INotificationManager>(ShinyHost.Resolve<INotificationManager>());`

That's worked just fine but we need to get access to some of our services within some Shiny delegates such as IPushDelegate. The approach I have taken is to expolicitly call Resolve on the FreshIOC container

`var breatheServices = FreshIOC.Container.Resolve<IBreatheServices>();
var clinicalService = FreshIOC.Container.Resolve<IClinicalService>();`

Under normal debugging and test scenarios that seems to work but recently I wired up Shiny Logging via the `ShinyStartup.ConfigureLogging` and that highlighted that sometimes those Resolve calls are failing. This makes me think I should be configure the dependencies needed in the Shiny delegates using the correct Shiny process.

I add some AddSingleton registration in ShinyStartup.ConfigureServices

public override void ConfigureServices(IServiceCollection services, IPlatform platform)
{
            services.UsePush<PushDelegate>();
            services.UseNotifications<NotificationDelegate>();

            services.AddSingleton<IDeviceInformationService, DeviceInformationService>();
            services.AddSingleton<INetworkInformationService, NetworkInformationService>();
            services.AddSingleton<IClinicalService, ClinicalService>();
            services.AddSingleton<IDataProviderService, DataProviderService>();
            services.AddSingleton<IBreatheServices, BreatheServices>();
}

And then added the dependencies to the PushDelagate ctor. Now when I run the app I get no errors but the PsuhDelegate seems to be invalid woith none of the breakpoints firing and no System.Diagnostics.Debug.WriteLine outputs appearing in the output window

1. Have I totally misunderstood how I should get my depedencies into the likes of the PushDelegate?
2. Is there a way of reusing the FreshMVVM IOC container with Shiny?

**Comment by aritchie** [comment]
Do you have a sample I could look at?  One thing you could do is resolve the PushDelegate in a try/catch and see what's invalid.  Shiny does usually log this for you.

**Reply by munkii** [reply]
Resolving PushDelegate was the key to seeing what was needed and once PushDelegate had it's dependencies it was being called  as expected again, Thanks.

The cascade of dependencies meant i needed 24 AddSingleton calls. Is there a way of sharing the FreshMVVM Container with Shiny?

**Reply by aritchie** [reply]
I have never used FreshMVVM so I'm not sure.  Does it support replacing the container or catch-all type when something isn't registered with it's container, so you could do something like.   freshMvvmContainer.ResolveMissing(type => ShinyHost.Resolve(type))?

**Reply by munkii** [reply]
What a can of worms!

TinyIOC does not support something like "ResolveMissing" however FreshMVVM does allow the complete replacement of it's IOC container.

One of the downsides of that are that all the handy ViewModel and View instantiations that Fresh did I now need to do myself (not then end of the word).

The showstopper looks like the interface I need to support to be able to replace the Container. Fresh wants to be able to Register stuff outside of the startup AND by name. I cannot see a way to Resolve or Register implementations with Shiny's container by name.

Does Shiny support a ResolveMissing approach? That way I could keep Fresh's IOC intact and then Resolve from that when Shiny didn't have something it needed.

**Reply by aritchie** [reply]
Shiny doesn't have it's own DI directly.  I build in some DI tricks, but nothing at the container level, so nothing that would help you with a "Resolve Missing" approach.  

You could try to wire all dependencies into shiny and expose the container shiny uses to Fresh?

**Reply by munkii** [reply]
>>You could try to wire all dependencies into shiny and expose the container shiny uses to Fresh"<<

That's what I have currently tried and it almost works. I have a basic version for just the Home page, implemented using IFreshIOC as mentioned here, https://github.com/rid00z/FreshMvvm#user-content-custom-ioc-containers

The trouble is that Fresh wants to be able to Register and Resolve by name and Shiny from what I can see does not allow me to do that.

**Reply by aritchie** [reply]
It really has nothing to do with Shiny.  I use Microsoft.Extensions.DependencyInjection because it has a good set of abstractions that are easy to replace.  Maybe there's a plugin to allow tinyioc into this layer?  

DryIoc supports missing resolves and named registration I believe.  Maybe you can look there?

**Reply by munkii** [reply]
This is all now working. 

- Leave FreshMVVM to look after all the View\ViewModel registration.
- Register all my services classes with Shiny on startup
- After startup I resolved the necessary VM dependencies into FreshMVVM by passing the instance from Shiny Resolve

i.e.

`this.Register<IRemindersService>(ShinyHost.Resolve<IRemindersService>());`



Sorry should have replied sooner.
